Preliminary stuff for herwig

Cvico opened the pull request:
Hi,

This is just a preliminary version of herwig config files for central herwig samples. There's no particular rush I think in having this implement in a short timescale, but I'm leaving the PR open just to have this in mind in the future. This might take a bit of time because if I'm not mistaken, it would require a bit of development in the backend.

I think there are a few things to discuss:

- **ExternalProducer**: herwig requires to launch an script that merges lhe files before starting the shower step. This can be seen in the `    postGenerationCommand = cms.untracked.vstring('mergeLHE.py', '-i', 'thread*/cmsgrid_final.lhe', '-o', 'cmsgrid_final.lhe'),` option in `ExternalLHEProducer_Powheg-herwig7.dat`.
- **Fragment**: There are a few things to mention here.
  - Center of mass energy and processParameters (which for herwig are often specific commands of herwig) can be configured using the same placeholders as for pythia8.     
  - I haven't found out what is the option `configFiles = cms.vstring()` in the fragment. Seems to me that this is to allow people to configure herwig commands using an input file that is a ".in" format. See the **only** example in cmssw that does not set this to just an empty variable [here](https://github.com/cms-sw/cmssw/blob/245daa4d7ad8e1412084c2d66381edfd8835d7ad/GeneratorInterface/Herwig7Interface/python/Herwig7_Dummy_ConfigFile_cff.py#L10). This config points to [this file](https://github.com/cms-sw/cmssw/blob/245daa4d7ad8e1412084c2d66381edfd8835d7ad/GeneratorInterface/Herwig7Interface/test/LHC-Matchbox.in#L155).
- **Input process**. I think for herwig samples we should have the mindset that by default these are variations of our commonly pythia8 showered samples. I've create the example I'm working now within TOP to test once the backend is set to consider herwig showers.

Cvico commented:
One more comment. I'm not sure if maybe we can add a replaceable in the common ExternalLHEProducer-Powheg.dat file so that it can be used for this as well.

sihyunjeon commented on Fragments/Generator/ExternalLHEProducer_Powheg-herwig7.dat:
```diff
@@ -0,0 +1,12 @@
+    outputFile = cms.string('cmsgrid_final.lhe'),
+    scriptName = cms.FileInPath('GeneratorInterface/LHEInterface/data/run_generic_tarball_cvmfs.sh'),
+    generateConcurrently = cms.untracked.bool($concurrentLHEforPowheg)
+    postGenerationCommand = cms.untracked.vstring('mergeLHE.py', '-i', 'thread*/cmsgrid_final.lhe', '-o', 'cmsgrid_final.lhe'),
```
hmm i would suggest to add this with a replaceable instead of having a new data file in case it's needed

sihyunjeon commented on Fragments/PartonShower/powheg-herwig7.dat:
```diff
@@ -0,0 +1,39 @@
+from Configuration.Generator.Herwig7Settings.Herwig7CH3TuneSettings_cfi import *
+from Configuration.Generator.Herwig7Settings.Herwig7LHEPowhegSettings_cfi import *
+from Configuration.Generator.Herwig7Settings.Herwig7PSWeightsSettings_cfi import *
+from Configuration.Generator.Herwig7Settings.Herwig7_7p1SettingsFor7p2_cfi import *
```
not sure what this really does, are you aware of this?

Cvico replied:
Uhm that is loading the Herwig CH3 tuning right? Now what I'm not so sure is the `Herwig7_7p1SettingsFor7p2_cfi`, not sure if that's the latest.

sihyunjeon replied:
we should ask dominic

Dominic-Stafford replied:
Hi, just to add here, the `Herwig7_7p1SettingsFor7p2` set a number of UE settings which changed when going from Herwig 7.1 to 7.2 back to their 7.1 values, since this was the release used to make the CH tunes

agrohsje replied:
@Dominic-Stafford wouldn't it make sense to add this to the CH settings? It will work with the old versions where this was already default and still with the new ones.  We could make an entry in the fragment that these are not actual settings of the tune but default ones that have changed and where used to derive the tune. The current name is a bit confusing. I will still work with 7.3 and it is related to the tune not the general program, so people might only want it if they are using CH3.

Dominic-Stafford replied:
Yes, maybe this would make sense, though it might also cause confusion for people taking parameters from old cards if we remove it. Maybe first I should ask Emyr is there was a motivation for making it a separate block? For 7.3 this might need to be expanded as maybe there will be further parameters which should be set back to their 7.1 values

sihyunjeon commented on Fragments/PartonShower/powheg-herwig7.dat:
```diff
@@ -0,0 +1,39 @@
+
+from Configuration.Generator.Herwig7Settings.Herwig7LHECommonSettings_cfi import *
+from Configuration.Generator.Herwig7Settings.Herwig7StableParticlesForDetector_cfi import *
+from Configuration.Generator.Herwig7Settings.Herwig7CH3TuneSettings_cfi import *
```
we should make the tunes configurable

Cvico commented:
Hi, I've updated the PR following Dominic's mail (thanks a lot for the info!!)

I also modified the `ExternalLHEProducer_Powheg.dat` card to have a replaceable "$additionalCommands" that can should be changed by whatever we put in the .json (see the example file under `NewTests` folder in powheg). If nothing is provided, it should automatically remove the replaceable (I'll ask geovanny to see if this can be done, because the decission should be made in the backend).

agrohsje commented on Cards/Powheg/NewTests/TWminustoLNu2Q_powheg-herwig7/TWminustoLNu2Q_powheg-herwig7.json:
```diff
@@ -0,0 +1,14 @@
+{
+    "gridpack_submit": false,
+    "gridpack_path": "SingleT/ST_wtch_DR_slc7_amd64_gcc10_CMSSW_12_4_8_TWminustoLNu2Q_powheg-pythia8.tgz",
```
The fact that this is hard-coded here is just for testing ?

Dominic-Stafford commented:
Hi, sorry, it's occurred to me it would probably be good to make these samples using the Herwig7HadronizerFilter rather than the old Herwig7GeneratorFilter. This is the change that fixes the issue of the wrong lhe events being saved due to Herwig silently skipping events, and is more relevant for mlm and FxFx where a lot of events get skipped in merging, but I think even for POWHEG Herwig will in some rare conditions veto an event, and this only need happen once for all subsequent events to have the wrong LHE information, so we should probably use the HadronizerFilter to be safe. How soon are you planning on producing these samples? The change hasn't made it into 12_4 yet, so should I ask the gen conveners to ask ORP for a new release?

agrohsje commented:
Yes. Please do so. I think it is good to have this consistently in Run3.
